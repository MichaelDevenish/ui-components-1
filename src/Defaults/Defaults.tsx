import React from 'react'

const sassGlobals = require('../sass/globals.scss')

interface ProvidedDefaults {
  /** Breakpoints used for the grid layout */
  breakpoints: {
    small: number
    medium: number
    large: number
    xlarge: number
    xxlarge: number
  }
}

interface Defaults extends ProvidedDefaults {
  /** Anchor component used for clickable links */
  AnchorComponent?: any
}

const defaults: ProvidedDefaults = {
  breakpoints: {
    small: parseInt(sassGlobals['breakpoint-small']),
    medium: parseInt(sassGlobals['breakpoint-medium']),
    large: parseInt(sassGlobals['breakpoint-large']),
    xlarge: parseInt(sassGlobals['breakpoint-xlarge']),
    xxlarge: parseInt(sassGlobals['breakpoint-xxlarge'])
  }
}

const DefaultsContext: React.Context<Defaults> = React.createContext(defaults)

const DefaultsConsumer = DefaultsContext.Consumer

const DefaultsProvider = DefaultsContext.Provider

export {
  ProvidedDefaults,
  Defaults,
  DefaultsConsumer,
  DefaultsProvider,
  defaults
}
