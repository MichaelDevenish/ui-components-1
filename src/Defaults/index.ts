export { Defaults, DefaultsConsumer, DefaultsProvider } from './Defaults'
