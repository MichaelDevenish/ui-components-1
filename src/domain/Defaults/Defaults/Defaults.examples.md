#### DefaultsProvider

DefaultsProvider can be used to supply a list of default values to ui components.
This is especially useful for having default configuration for common things,
such as the component to use for anchors throughout the design system:

```
const { DefaultsProvider } = require('./Defaults');

<DefaultsProvider
  value={{
    AnchorComponent: 'i'
  }}
>
  <ActionLink
    href='#'
  >
    Your text here
  </ActionLink>
</DefaultsProvider>
```

#### DefaultsConsumer

DefaultsConsumer can be used within ui components to access the values given
in the defaults provider. This works cross repositories and through any
depth in the react dom tree.

```
const { DefaultsProvider, DefaultsConsumer } = require('./Defaults');

<DefaultsProvider
  value={{
    coolKey: 'coolio'
  }}
>
  <div>
    <DefaultsConsumer>
      {defaultValues => <b>{defaultValues.coolKey}</b>}
    </DefaultsConsumer>
  </div>
</DefaultsProvider>
```
