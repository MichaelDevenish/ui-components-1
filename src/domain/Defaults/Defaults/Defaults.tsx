import React from 'react'

import { GridProvider } from '../../Grids'

interface IDefaults {
  /** Anchor component used for clickable links */
  tenorApiKey?: string
  AnchorComponent?: React.ComponentType<any>
}

interface IDefaultsProviders {
  value?: Partial<IDefaults>
  children?: React.ReactNode
}

const defaults: IDefaults = {}

const DefaultsContext: React.Context<IDefaults> = React.createContext(defaults)

const DefaultsConsumer = DefaultsContext.Consumer

class DefaultsProvider extends React.PureComponent<IDefaultsProviders> {
  public render(): JSX.Element {
    const {
      value,
      children
    } = this.props

    return (
      <DefaultsContext.Provider
        value={{
          ...defaults,
          ...value
        }}
      >
        {children}
      </DefaultsContext.Provider>
    )
  }
}

export {
  IDefaults,
  IDefaultsProviders,
  DefaultsConsumer,
  DefaultsProvider,
  DefaultsContext,
  defaults
}
