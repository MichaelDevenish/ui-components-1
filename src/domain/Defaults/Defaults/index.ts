export { defaults, IDefaults, DefaultsConsumer, DefaultsProvider, DefaultsContext } from './Defaults'
