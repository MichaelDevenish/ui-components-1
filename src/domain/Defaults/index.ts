export { defaults, IDefaults, DefaultsConsumer, DefaultsProvider, DefaultsContext } from './Defaults'
export { IWithDefaults, withDefaults } from './withDefaults'
