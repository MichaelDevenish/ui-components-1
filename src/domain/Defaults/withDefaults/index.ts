export { IWithDefaults, withDefaults } from './withDefaults'
