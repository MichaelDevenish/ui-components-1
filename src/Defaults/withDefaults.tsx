import React from 'react'
import { Subtract } from 'utility-types'
import { Defaults, DefaultsConsumer } from './Defaults'

interface IWithDefaults {
  defaults: Defaults
}

const withDefaults = <P extends IWithDefaults>(DefaultsAwareComponent: React.ComponentType<P>) => (
  class extends React.PureComponent<Subtract<P, IWithDefaults>> {
    public render (): JSX.Element {
      return (
        <DefaultsConsumer>
          {(defaults: Defaults) =>
            <DefaultsAwareComponent
              {...this.props}
              defaults={defaults}
            />
          }
        </DefaultsConsumer>
      )
    }
  }
)

export {
  IWithDefaults,
  withDefaults
}
