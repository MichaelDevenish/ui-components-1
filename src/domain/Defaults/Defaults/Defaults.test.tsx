import React from 'react'
import { mount } from 'enzyme'
import { DefaultsProvider, DefaultsConsumer } from './Defaults'

describe('<Defaults />', () => {
  it('should provide default values', () => {
    const mockFn = jest.fn()

    const wrapper = mount(
      <DefaultsProvider>
        <DefaultsConsumer>
          {defaults => mockFn(defaults)}
        </DefaultsConsumer>
      </DefaultsProvider>
    )

    expect(mockFn.mock.calls).toMatchSnapshot()
  })

  it('should provide values with defaults', () => {
    const mockFn = jest.fn()

    const wrapper = mount(
      <DefaultsProvider
        value={{
          AnchorComponent: () => <div />
        }}
      >
        <DefaultsConsumer>
          {defaults => mockFn(defaults)}
        </DefaultsConsumer>
      </DefaultsProvider>
    )

    expect(mockFn.mock.calls).toMatchSnapshot()
  })

  it('should overwrite default values', () => {
    const mockFn = jest.fn()

    const wrapper = mount(
      <DefaultsProvider
        value={{
          breakpoints: {
            xsmall: 0,
            small: 100,
            medium: 200,
            large: 300,
            xlarge: 400
          }
        }}
      >
        <DefaultsConsumer>
          {defaults => mockFn(defaults)}
        </DefaultsConsumer>
      </DefaultsProvider>
    )

    expect(mockFn.mock.calls).toMatchSnapshot()
  })

  it('should still provide default values without Provider', () => {
    const mockFn = jest.fn()

    const wrapper = mount(
      <DefaultsConsumer>
        {defaults => mockFn(defaults)}
      </DefaultsConsumer>
    )

    expect(mockFn.mock.calls).toMatchSnapshot()
  })
})
