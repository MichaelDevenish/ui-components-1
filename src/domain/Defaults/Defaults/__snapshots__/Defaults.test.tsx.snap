// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Defaults /> should overwrite default values 1`] = `
Array [
  Array [
    Object {
      "breakpoints": Object {
        "large": 300,
        "medium": 200,
        "small": 100,
        "xlarge": 400,
        "xsmall": 0,
      },
    },
  ],
]
`;

exports[`<Defaults /> should provide default values 1`] = `
Array [
  Array [
    Object {
      "breakpoints": Object {
        "large": 1024,
        "medium": 640,
        "small": 0,
        "xlarge": 1200,
        "xxlarge": 1440,
      },
    },
  ],
]
`;

exports[`<Defaults /> should provide values with defaults 1`] = `
Array [
  Array [
    Object {
      "AnchorComponent": [Function],
      "breakpoints": Object {
        "large": 1024,
        "medium": 640,
        "small": 0,
        "xlarge": 1200,
        "xxlarge": 1440,
      },
    },
  ],
]
`;

exports[`<Defaults /> should still provide default values without Provider 1`] = `
Array [
  Array [
    Object {
      "breakpoints": Object {
        "large": 1024,
        "medium": 640,
        "small": 0,
        "xlarge": 1200,
        "xxlarge": 1440,
      },
    },
  ],
]
`;
